Rewrite signed elements structure

tvdijen opened the pull request.

tvdijen committed:
Rewrite signed elements structure

tvdijen committed:
Add toUnsignedXML to the SignableElementTrait

tvdijen committed:
Any signed element should have a signature set

tvdijen committed:
Further finalize abstract

tvdijen committed:
Add testable CustomSignable and CustomSigned

tvdijen committed:
Test signable

jaimeperez commented on src/XML/SignableElementTrait.php:
```diff
@@ -0,0 +1,44 @@
+     * @return \DOMElement The signed element.
+     * @throws \Exception If an error occurs while trying to sign.
+     */
+    private function toSignedXML(XMLSecurityKey $signingKey, array $certificates, DOMNode $insertBefore = null): DOMElement
```
I think this method is not really necessary. A _signable_ element must implement the `sign()` method, which will in turn return a _signed_ element. If we want the actual DOM representation of the signed element, we would only need to call `toXML()` on the `SignedElement` we just obtained.

tvdijen replied:
Agreed, I'll try and kill it

jaimeperez commented on src/XML/SignedElementInterface.php:
```diff
@@ -0,0 +1,35 @@
+     * @param \SimpleSAML\XMLSecurity\XMLSecurityKey $key The key we should check against.
+     * @return bool True if successful, false if we don't have a signature that can be verified.
+     */
+    public function validate(XMLSecurityKey $key): bool;
```
I think, in order to encourage good practices, and also to avoid any possible implementation errors (e.g. we have a signed element, and we just call `getElement()` to get the unsigned element without validating the signature first), the `validate()` method should return the actual element whose signature has been verified.

That would mean that when the signature validation fails, we should throw an exception (which is a good way to ensure the failure is noticeable, I believe), instead of returning false. It would also prevent us from having any processing security issue. We'll need support from xmlseclibs or our fork though to get this done, since verify should call `fromXML()` on the target class (the signable element), passing the actual, canonicalised an transformed DOM (not the original), and as far as I recall, that's not possible with xmlseclibs.

jaimeperez commented on src/XML/SignedElementTrait.php:
```diff
@@ -0,0 +1,108 @@
+    public function getValidatingCertificates(): array
+    {
+        $ret = [];
+        foreach ($this->signature->getCertificates() as $cert) {
```
I think this can be quite expensive. Would this be an alternative approach?

1. Require `validate()` before calling `getValidatingCertificates()`. If we call the latter first, raise an exception.
2. During `validate()`, if the signature validation succeeds with the given key, check if it corresponds to any of the certificates stored internally. Build a list of validating certificates with all where the passed key corresponds to them.
3. When calling `getValidatingCertificates()` after `validate()`, return the list we just built. If none of the certificates  in the signature correspond to the given key, or no certificates where found, we just return an empty list.

jaimeperez commented on tests/XML/CustomSigned.php:
```diff
@@ -0,0 +1,54 @@
+     * @param \DOMElement $xml
+     * @return self
+     */
+    public static function fromXML(DOMElement $xml): object
```
I feel like this method could be implemented generically. We don't really need to check the name of the element, since the _Signed*_ convention is just representing a signed element, but not an actual element named like that. That check will be done later when `fromXML()` is called on the actual class (e.g. `Assertion`). So then we just need to create the `Signed*` object by calling the constructor, passing the signature and the DOM. Something like this:

```php
    public static function fromXML(DOMElement $xml): SignedElementInterface
    {
        $signature = Signature::getChildrenOfClass($xml);
        Assert::minCount($signature, 1, MissingElementException::class);
        Assert::maxCount($signature, 1, TooManyElementsException::class);

        return new self($xml, $signature);
    }
```

If we make the `validate()` method return the `SignableElementInterface` object, then we don't even need to create that object first. We can just postpone the creation of the object until validation, which is also safer, I think.

jaimeperez commented on tests/XML/CustomSignableTest.php:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * Test serialization / unserialization
+     */
+    public function testSerialization(): void
```
Serialisation should be tested on `SignedElementInterface` objects, not on the _signable_ ones, since those are covered already by extending the `AbstractXMLElement` class. The signed ones are the problematic, since they keep the DOM in a property.

tvdijen replied:
How can I run tests on an interface?  I created this real-life dummy class to be able to test it..

jaimeperez replied:
I ended up resolving my own comment here by accident 😅

jaimeperez commented on tests/XML/CustomSigned.php:
```diff
@@ -0,0 +1,54 @@
+//        Assert::minCount($signatureElement, 1, TooManyElementsException::class);
+        $signature = Signature::getChildrenOfClass($xml);
+        Assert::minCount($signature, 1, MissingElementException::class);
+        Assert::minCount($signature, 1, TooManyElementsException::class);
```
`Assert::maxCount()` here maybe? 😉

jaimeperez commented on src/XML/AbstractSignedXMLElement.php:
```diff
@@ -0,0 +1,123 @@
+     * @param \SimpleSAML\XMLSecurity\XML\SignableElementInterface $elt
+     * @param \SimpleSAML\XMLSecurity\XML\ds\Signature $signature
+     */
+    protected function __construct(DOMElement $xml, SignableElementInterface $elt, Signature $signature)
```
As I mentioned in another comment, I would remove the actual element from the constructor, so that it is only built after verifying the signature.

jaimeperez commented on src/XML/AbstractSignedXMLElement.php:
```diff
@@ -0,0 +1,123 @@
+     *
+     * @return \DOMElement
+     */
+    public function getStructure(): DOMElement
```
I think the getter here is redundant. `toXML()` does the same, and gives as a uniform interface across the library.

jaimeperez commented on src/XML/AbstractSignedXMLElement.php:
```diff
@@ -0,0 +1,123 @@
+     *
+     * @param \DOMElement $structure
+     */
+    private function setStructure(DOMElement $structure): void
```
And here, since there's no validation at all on the DOM element, I think we can just skip the setter as well.

tvdijen replied:
Agreed

jaimeperez commented on src/XML/SignedElementInterface.php:
```diff
@@ -0,0 +1,35 @@
+
+
+    /**
+     * Validate this element against a public key.
```
This won't necessarily be a public key.

tvdijen replied:
What else would it be?

tvdijen committed:
Remove getter/setter for element

tvdijen committed:
Remove unnecessary SignableElementTrait

tvdijen committed:
Fix

tvdijen committed:
Fix

tvdijen committed:
Make SignedElement*::fromXML generic

tvdijen committed:
Change parent

tvdijen committed:
Use SerializableXMLTestTrait

tvdijen committed:
Return unsigned element

tvdijen committed:
Remove getStructure and use toXML instead

tvdijen committed:
Fix class-name

tvdijen committed:
Fix CustomSignedTest

tvdijen committed:
Fix return value

tvdijen committed:
Mess up

tvdijen committed:
Fix

tvdijen committed:
Leave the Chunk in the trunk

tvdijen committed:
Make Signed classes serializable

tvdijen committed:
s/self/static

tvdijen committed:
Improve tests

jaimeperez committed:
Make sure to ignore extra whitespaces in InclusiveNamespaces@PrefixList

jaimeperez committed:
Minor fixes to ds:XPath implementation

tvdijen commented on src/XML/ec/InclusiveNamespaces.php:
```diff
@@ -62,7 +62,7 @@ public static function fromXML(DOMElement $xml): object
         $prefixes = self::getAttribute($xml, 'PrefixList', '');
 
-        return new self(explode(' ', $prefixes));
+        return new self(array_filter(explode(' ', $prefixes), function ($v, $k) { return !empty($v); }));
```
From the PHP-docs:   "If no callback is supplied, all empty entries of array will be removed. See empty() for how PHP defines empty in this case."

jaimeperez replied:
Good catch 😉 

I thought about leaving it off as well, but on a second thought, explicit is better, since then we know it will work even if for whatever reason `' '` stops being evaluated to `false` in the future.

jaimeperez committed:
Remove closure from InclusiveNamespaces

jaimeperez committed:
Nitpicks

Add strict_types declarations and modify package names.

jaimeperez committed:
Make SignatureAlgorithmFactory reusable for custom algs

This commit rewrites the way the signature algorithms work and are created via the factory. It offloads the responsibility to check what algorithms they support to the signer implementations themselves, and makes them return their supported algs.

It also introduces a mechanism in the factory so that you can register new signer implementations, and use them automatically in your code.

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

Update tests for SignatureAlgorithmFactory

jaimeperez committed:
Move the hash() function to our Security utility

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Add new XML utility class

For now, move there the canonicalization and transform processing mechanisms.

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
First attempt to implement signing in XML objects

This introduces a SignedElementTrait that gives us most of what we need.

The SignableElementInterface::sign() method is now changed to receive a SignatureAlgorithm, the canonicalization algorithm and an optional KeyInfo element to add to the resulting signature. It also introduces a new method getId() that each element must implement to return their own ID used by the signature reference.

The new trait implements the sign() method of the interface, and also provides a doSign() method that objects can use to effectively sign them. Calling it will generate a Signature object and keep it in a property. This object can then be used in toXML() in the object and inserted wherever we want, using the insertBefore() convenience method (also provided by the trait).

The CustomSignable class demonstrates how to use the functionality. CustomSigned and its corresponding test disappears (as it made no sense in this context, since the actual XML elements are called CustomSignable).

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Use the local XML-related constants

Downgrade xml-common to 0.7.0, use the local constants.

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

tvdijen commented on src/Alg/Signature/HMAC.php:
```diff
@@ -24,10 +24,26 @@ class HMAC extends AbstractSigner implements SignatureAlgorithm
+    public static function getSupportedAlgorithms(): array
+    {
+        return [
+            C::SIG_HMAC_SHA1,
```
Should we create a constant array, like we do for like digest algorithms?

tvdijen commented on src/Alg/Signature/RSA.php:
```diff
@@ -24,10 +24,26 @@ class RSA extends AbstractSigner implements SignatureAlgorithm
+    public static function getSupportedAlgorithms(): array
+    {
+        return [
+            C::SIG_RSA_RIPEMD160,
```
Same

tvdijen commented on src/Constants.php:
```diff
@@ -107,6 +107,24 @@ class Constants extends \SimpleSAML\XML\Constants
     public const SIG_HMAC_SHA512 = 'http://www.w3.org/2001/04/xmldsig-more#hmac-sha512';
     public const SIG_HMAC_RIPEMD160 = 'http://www.w3.org/2001/04/xmldsig-more#hmac-ripemd160';
 
+    public static $RSA_DIGESTS = [
```
Seems like you did create the static arrays, but you're not using them

jaimeperez replied:
Actually, I did that on purpose, so that it would be absolutely explicit what we support per algorithm. But now that you have changed it it's fine too, not a big deal...

tvdijen commented on src/Constants.php:
```diff
@@ -128,11 +128,11 @@ class Constants extends \SimpleSAML\XML\Constants
      */
-    public const XMLDSIGNS = 'http://www.w3.org/2000/09/xmldsig#';
-    public const XMLDSIG11NS = 'http://www.w3.org/2009/xmldsig11#';
+    public const NS_XDSIG = 'http://www.w3.org/2000/09/xmldsig#';
```
Instead of renaming the constants, you should have probably renamed the cases where we use them..

tvdijen replied:
As discussed, we will start using new NS_ prefixed names.

jaimeperez committed:
Add CanonicalizableInterface

This interface provides a method to canonicalise elements implementing it. This allows us to be sure that the actual input for the C14N algorithm is the same as the original, if the object was created with fromXML().

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

tvdijen commented on src/XML/CanonicalizableInterface.php:
```diff
@@ -0,0 +1,36 @@
+ * by this interface.
+ *
+ * If the original DOM object is retained, please remember to implement the \Serializable interface so that the
+ * result of serialisation is the original string representation of the object, since DOM objects are not serializable.
```
SerialiZation

jaimeperez replied:
Excuse me Sir, I just happen to be taught British English 😝

tvdijen commented on src/XML/CanonicalizableInterface.php:
```diff
@@ -0,0 +1,36 @@
+use SimpleSAML\XML\XMLElementInterface;
+
+/**
+ * An interface for objects that can be canonicalised.
```
CanonicaliZed

tvdijen commented on src/XML/CanonicalizableTrait.php:
```diff
@@ -0,0 +1,58 @@
+ *
+ * @package simplesamlphp/xml-security
+ */
+trait CanonicalizableTrait
```
Regarding the name of the trait and the interface....  Canonicalizable-what?
Maybe rename to CanonicalizableXMLTrait / CanonicalizableXMLInterface or CanonicalizableElementTrait / CanonicalizableElementInterface

jaimeperez replied:
I saw this coming 😉 

Yeah, I already thought about `CanonicalizableElement*`, but since it was quite long I left it like that. I'll rename it 👍

tvdijen commented on src/XML/ds/SignedInfo.php:
```diff
@@ -34,6 +37,11 @@ final class SignedInfo extends AbstractDsElement
+
 
     /**
      * Initialize a SignedIfno.
```
Let's fix this typo while at it

tvdijen committed:
Use array constants

tvdijen committed:
Fix typos

tvdijen committed:
Fix booboo

tvdijen committed:
Bump xml-common

tvdijen committed:
Unrelated; cleanup

tvdijen committed:
Add CipherValue XML-class

tvdijen committed:
Implement CipherValue XML element

tvdijen committed:
Add CarriedKeyName XML-class

jaimeperez committed:
Rename Canonicalize* interface and trait

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Bugfix

What we need to return in the getSupportedAlgorithms() method is the list of signature algorithms, not their corresponding digests.

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

tvdijen committed:
Implement CarriedKeyName XML-class

tvdijen committed:
Add missing @covers

Raise coverage

tvdijen committed:
Remove overhead

tvdijen committed:
Start cleaning up old code with hooks to xmlseclibs

tvdijen commented on src/Utils/XML.php:
```diff
@@ -0,0 +1,144 @@
+                        $arXPath = [];
+                        $arXPath['query'] = '(.//. | .//@* | .//namespace::*)[' . $xpath->getExpression() . ']';
+                        $arXpath['namespaces'] = $xpath->getNamespaces();
+                        // TODO: review if $nsnode->localName is equivalent to the keys in getNamespaces()
```
Was the code not working properly?

tvdijen committed:
Default to empty string;  Scrutinizer warns us to do additional type check now, but the default value is never returned anyway

tvdijen committed:
Add constants for ASN tags and sizes to improve readability

tvdijen committed:
Use assertion instead of returning null for string out-of-bounds

tvdijen committed:
Silence Scrutinizer;  the syntax is just fine, but Scrutinizer doesn't like the alternative syntax

tvdijen committed:
Fix implicit array to bool cast

tvdijen committed:
Remove array test;  the properties are always an array

tvdijen committed:
Fix phpdoc

tvdijen committed:
Fix parentheses

tvdijen committed:
Fix incorrect return type

tvdijen committed:
Fix phpdoc

jaimeperez committed:
Add a method to SignatureAlgorithm to get the key

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Add constants for X509 certificate header/footer

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Use our own Assert in SignableElementTrait

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Phpdoc fixes

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Make SignableElementInterface extend CanonicalizableElementInterface

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

tvdijen committed:
Fix DOMXPath use

tvdijen committed:
Add use-statements for builtins

jaimeperez committed:
Minor fixes

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Add a method to ds:Reference to determine if it's an XPointer

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Add more signed documents

These new documents can be used in tests to verify different features.

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Signable elements should call toXML()

We got rid of the initial extra toSignedXML() method. If we want to sign an element, we should just call sign() on it, then get the result with toXML().

This means it will be toXML() the method that actually does the signing, while sign() will simply prepare everything for it and signal that the element should be signed when calling toXML().

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Use short name for Constants class

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Fix wrong package name

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
No longer override c14n alg

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

jaimeperez committed:
Implement signature verification, refactor signing

This commit adds support for verifying signatures (and renames the API method to verify(), which is the correct term we should us in this context, instead of validate()).

It also refactors the way we sign, and adds support for canonicalization algorithms with comments, which were in practice not supported before.

Signed-off-by: Jaime Pérez <jaimepc@gmail.com>

tvdijen committed:
Fix nits

tvdijen committed:
Fix nits

tvdijen committed:
Fix file name

tvdijen committed:
Fix phpcs issues

tvdijen committed:
Fix psalm issue
